pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -607,35 +609,89 @@ protected void handlePartitionMetadataRequest(CommandPartitionedTopicMetadata pa
+                    getBrokerService().isAllowAutoTopicCreationAsync(topicName).thenAccept(brokerAllowAutoCreate -> {
+                        boolean autoCreateIfNotExist = partitionMetadata.isMetadataAutoCreationEnabled();
+                        if (!autoCreateIfNotExist) {
+                            final NamespaceResources namespaceResources = getBrokerService().pulsar()
+                                    .getPulsarResources().getNamespaceResources();
+                            final TopicResources topicResources = getBrokerService().pulsar().getPulsarResources()
+                                    .getTopicResources();
+                            namespaceResources.getPartitionedTopicResources()
+                                .getPartitionedTopicMetadataAsync(topicName, false)
+                                .thenAccept(metadata -> {
+                                    if (metadata.isPresent()) {
+                                        commandSender.sendPartitionMetadataResponse(metadata.get().partitions,
+                                                requestId);
+                                        lookupSemaphore.release();
+                                        return;
+                                    }
+                                    if (topicName.isPersistent()) {
+                                        topicResources.persistentTopicExists(topicName).thenAccept(exists -> {
+                                            if (exists) {
+                                                commandSender.sendPartitionMetadataResponse(0, requestId);
+                                                lookupSemaphore.release();
+                                                return;
+                                            }
+                                            writeAndFlush(Commands.newPartitionMetadataResponse(
+                                                    ServerError.TopicNotFound, "", requestId));
+                                            lookupSemaphore.release();
+                                            return;
+                                        }).exceptionally(ex -> {
+                                            log.error("{} {} Failed to get partition metadata", topicName,
+                                                    ServerCnx.this.toString(), ex);
+                                            writeAndFlush(
+                                                    Commands.newPartitionMetadataResponse(ServerError.MetadataError,
+                                                            "Failed to check partition metadata",
+                                                            requestId));
+                                            lookupSemaphore.release();
+                                            return null;
+                                        });
+                                    }
```

Comment:
Do we need to release the `lookupSemaphore` when handling a non-persistent topic?

Reply:
yes， it's better to release the semaphore  using a final `thenAccept`

Reply:
Fixed. And added the tests to cover the following scenarios:
- `non-persistent topic`
- `semaphore`